pkg/virt-launcher/monitor.go

```diff
@@ -182,6 +183,11 @@ func (mon *monitor) monitorLoop(startTimeout time.Duration, signalStopChan chan
 
 			mon.gracefulShutdownCallback()
 			mon.gracePeriodStartTime = time.Now().UTC().Unix()
+		default:
+			if cmdserver.ReceivedEarlyExitSignal() {
+				log.Log.Infof("received early exit signal")
+				mon.isDone = true
+			}
```

Comment:
the early exit signal was only meant as an optimization. This optimization was added when we prevented migrations from occurring until any previous target pods (from a failed migration) had completely terminated. Without this signal, there were situations where the migration failed so quickly that the domain never came online, then we'd have to wait the qemu timeout (~ 5 minutes) before the migration could be successfully triggered again.


Once the domain is alive, we expect Libvirt to keep the domain consistent for us during the live migration. If the migration fails, then the expectation is that libvirt on the target node tears down the qemu process, never allowing it to exit a "paused" state... that assumes pre-copy live migration. I'd expect the something similar with post-copy. Libvirt should only allow the domain to be alive and actively running in one location. 

If these assumptions are not holding true, then this is likely a libvirt bug

Reply:
Yes that sounds really bad. Can we determine what exactly said that the migration failed? Was it our code? Was it libvirt? Isecond @davidvossel  here. If the second one, we should probably add a bug report to libvirt. If the first one, we should probably re-evaluate our current design. This should be "save" at any time in the sense that the target can never start running. Just killing it is not good enough.